File: genetic-packer/src/main/java/genetic/packer/mapper/CellCloningMapper.java
package genetic.packer.mapper;

import java.util.function.Function;

import genetic.api.individual.Cell;
import genetic.api.individual.CellBuilder;
import javafx.scene.shape.Box;
import org.mapstruct.Mapper;

/**
 * @author piotr.larysz
 */
@Mapper
public abstract class CellCloningMapper implements Function<Cell<Box>, Cell<Box>> {

    @Override
    public Cell<Box> apply(Cell<Box> cell) {
        final Box box = cell.getNucleus();
        final Box clonedBox = new Box(box.getWidth(), box.getHeight(), box.getDepth());
        clonedBox.setTranslateX(box.getTranslateX());
        clonedBox.setTranslateY(box.getTranslateY());
        clonedBox.setTranslateZ(box.getTranslateZ());
        return new CellBuilder<Box>().nucleus(clonedBox)
            .processingOrder(cell.getProcessingOrder())
            .order(cell.getOrder())
            .build();
    }
}

File: genetic-packer/src/main/java/genetic/packer/mapper/CellCloningMapper.kt
package genetic.packer.mapper

import genetic.api.individual.Cell
import javafx.scene.shape.Box

abstract class CellCloningMapper : (Cell<Box>) -> Cell<Box> {

  override fun invoke(cell: Cell<Box>): Cell<Box> {
    val box = cell.nucleus
    val clonedBox = Box(box.width, box.height, box.depth)
    clonedBox.translateX = box.translateX
    clonedBox.translateY = box.translateY
    clonedBox.translateZ = box.translateZ
    return Cell(
        nucleus = clonedBox,
        processingOrder = cell.processingOrder,
        order = cell.order
    )
  }
}